### ISSUE TITLE
VMCluster status is always "expanding"

### ISSUE THREAD

**Issue by vozerov** [VictoriaMetrics/operator#677]
Hey everyone!

Got some error with Victoria Metrics operator. I am using latest version (victoriametrics/operator:v0.34.1). 

My VMCluster looks like:

```sh
apiVersion: operator.victoriametrics.com/v1beta1
kind: VMCluster
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"operator.victoriametrics.com/v1beta1","kind":"VMCluster","metadata":{"annotations":{},"labels":{"app.kubernetes.io/instance":"vm","app.kubernetes.io/managed-by":"Helm","app.kubernetes.io/name":"victoria-metrics-k8s-stack","app.kubernetes.io/version":"v1.91.2","helm.sh/chart":"victoria-metrics-k8s-stack-0.16.3"},"name":"vm-victoria-metrics-k8s-stack","namespace":"p-platform"},"spec":{"imagePullSecrets":[],"replicationFactor":1,"retentionPeriod":"14","vminsert":{"extraArgs":{},"image":{"tag":"v1.91.2-cluster"},"nodeSelector":{"usecase":"system"},"replicaCount":2,"resources":{}},"vmselect":{"cacheMountPath":"/select-cache","extraArgs":{"vmalert.proxyURL":"http://vmalert-vm-victoria-metrics-k8s-stack.p-platform.svc:8080"},"image":{"tag":"v1.91.2-cluster"},"nodeSelector":{"usecase":"system"},"replicaCount":2,"resources":{},"storage":{"volumeClaimTemplate":{"spec":{"resources":{"requests":{"storage":"10Gi"}}}}}},"vmstorage":{"image":{"tag":"v1.91.2-cluster"},"nodeSelector":{"usecase":"system"},"replicaCount":2,"resources":{},"storage":{"volumeClaimTemplate":{"spec":{"resources":{"requests":{"storage":"50Gi"}}}}},"storageDataPath":"/vm-data"}}}
    operator.victoriametrics/last-applied-spec: '{"retentionPeriod":"14","replicationFactor":1,"vmselect":{"image":{"tag":"v1.91.2-cluster"},"replicaCount":3,"resources":{},"cacheMountPath":"/select-cache","storage":{"volumeClaimTemplate":{"metadata":{},"spec":{"resources":{"requests":{"storage":"10Gi"}}},"status":{}}},"extraArgs":{"vmalert.proxyURL":"http://vmalert-vm-victoria-metrics-k8s-stack.p-platform.svc:8080"},"nodeSelector":{"usecase":"system"}},"vminsert":{"image":{"tag":"v1.91.2-cluster"},"replicaCount":2,"resources":{},"nodeSelector":{"usecase":"system"}},"vmstorage":{"image":{"tag":"v1.91.2-cluster"},"replicaCount":2,"resources":{},"storageDataPath":"/vm-data","storage":{"volumeClaimTemplate":{"metadata":{},"spec":{"resources":{"requests":{"storage":"50Gi"}}},"status":{}}},"nodeSelector":{"usecase":"system"}}}'
  creationTimestamp: "2023-06-22T23:19:31Z"
  finalizers:
  - apps.victoriametrics.com/finalizer
  generation: 6
  labels:
    app.kubernetes.io/instance: vm
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: victoria-metrics-k8s-stack
    app.kubernetes.io/version: v1.91.2
    helm.sh/chart: victoria-metrics-k8s-stack-0.16.3
  name: vm-victoria-metrics-k8s-stack
  namespace: p-platform
  resourceVersion: "3337023"
  uid: e8a18ef2-62f0-4604-846f-1e0e3b2f3236
spec:
  imagePullSecrets: []
  replicationFactor: 1
  retentionPeriod: "14"
  vminsert:
    extraArgs: {}
    image:
      tag: v1.91.2-cluster
    nodeSelector:
      usecase: system
    replicaCount: 2
    resources: {}
  vmselect:
    cacheMountPath: /select-cache
    extraArgs:
      vmalert.proxyURL: http://vmalert-vm-victoria-metrics-k8s-stack.p-platform.svc:8080
    image:
      tag: v1.91.2-cluster
    nodeSelector:
      usecase: system
    replicaCount: 3
    resources: {}
    storage:
      volumeClaimTemplate:
        metadata: {}
        spec:
          resources:
            requests:
              storage: 10Gi
        status: {}
  vmstorage:
    image:
      tag: v1.91.2-cluster
    nodeSelector:
      usecase: system
    replicaCount: 2
    resources: {}
    storage:
      volumeClaimTemplate:
        metadata: {}
        spec:
          resources:
            requests:
              storage: 50Gi
        status: {}
    storageDataPath: /vm-data
status:
  clusterStatus: expanding
  updateFailCount: 0
```

Get VMCluster:

```sh
$ kubectl get vmcluster -o wide
NAME                            INSERT COUNT   STORAGE COUNT   SELECT COUNT   AGE    STATUS
vm-victoria-metrics-k8s-stack   2              2               3              120m   expanding
```

The problem is that all pods have been already created:

```sh
$ kubectl get pods
vm-kube-state-metrics-6654db7c44-ht6ns                    1/1     Running   0          123m
vm-prometheus-node-exporter-22kct                         1/1     Running   0          123m
vm-prometheus-node-exporter-f8mph                         1/1     Running   0          123m
vm-prometheus-node-exporter-fzlwc                         1/1     Running   0          123m
vm-prometheus-node-exporter-ld8q2                         1/1     Running   0          123m
vm-prometheus-node-exporter-x72hr                         1/1     Running   0          123m
vm-victoria-metrics-operator-59b9f69b88-6gt2b             1/1     Running   0          5s
vmagent-vm-victoria-metrics-k8s-stack-86598f5b54-hj2rj    2/2     Running   0          121m
vmalert-vm-victoria-metrics-k8s-stack-5b7989fcd7-mx7b4    2/2     Running   0          121m
vmalertmanager-vm-victoria-metrics-k8s-stack-0            2/2     Running   0          121m
vminsert-vm-victoria-metrics-k8s-stack-5997b64d9b-2rw5r   1/1     Running   0          121m
vminsert-vm-victoria-metrics-k8s-stack-5997b64d9b-scmcq   1/1     Running   0          121m
vmselect-vm-victoria-metrics-k8s-stack-0                  1/1     Running   0          5m25s
vmselect-vm-victoria-metrics-k8s-stack-1                  1/1     Running   0          4m49s
vmselect-vm-victoria-metrics-k8s-stack-2                  1/1     Running   0          5m58s
vmstorage-vm-victoria-metrics-k8s-stack-0                 1/1     Running   0          121m
vmstorage-vm-victoria-metrics-k8s-stack-1                 1/1     Running   0          121m
```

So the cluster is working and all components are up and running. But the status is still "expanding". I've checked operator's logs and found:

```sh
{"level":"info","ts":1687483018.2646513,"logger":"controller","msg":"Reconciling VMCluster","Request.Namespace":"p-platform","Request.Name":"vm-victoria-metrics-k8s-stack"}
{"level":"error","ts":1687483018.2755868,"msg":"Reconciler error","controller":"vmcluster","controllerGroup":"operator.victoriametrics.com","controllerKind":"VMCluster","VMCluster":{"name":"vm-victoria-metrics-k8s-stack","namespace":"p-platform"},"namespace":"p-platform","name":"vm-victoria-metrics-k8s-stack","reconcileID":"eac5c320-f0ee-47da-8caa-8e1df6096336","error":"cannot set expanding status for cluster: Operation cannot be fulfilled on vmclusters.operator.victoriametrics.com \"vm-victoria-metrics-k8s-stack\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:326\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234"}
```

Don't know what is going on here, but i've checked vmcluster resource with kubectl and diff and found that resourceVersion is constantly changing (every second or so). But i don't think it's related to the issue above.

Thank you!

**Comment by Haleygo** [comment]
Hello! 
Is there any other operator in your cluster may update the vmcluster? Can you check the managedFields of vmcluster by `kubectl get vmcluster -oyaml --show-managed-fields=true`, and see the latest one.
>but i've checked vmcluster resource with kubectl and diff and found that resourceVersion is constantly changing (every second or so).
If you stop this operator, will the vmcluster still be updated every second? If so, who is the manager.

**Comment by vozerov** [comment]
@Haleygo Hello, thanks for the fast response.

Just double-checked - exactly one vm operator is running in the cluster across all namespaces.

Latest managed-field:

```sh
    - apiVersion: operator.victoriametrics.com/v1beta1
      fieldsType: FieldsV1
      fieldsV1:
        f:status:
          .: {}
          f:clusterStatus: {}
          f:updateFailCount: {}
      manager: manager
      operation: Update
      subresource: status
      time: "2023-06-23T01:43:24Z"
```

**Comment by Haleygo** [comment]
Looks like a bug then. 
Did you update vmselect replica to 2 using kubectl after vmcluster created?

**Comment by vozerov** [comment]
@Haleygo I did the following:

1. Used "vm/victoria-metrics-k8s-stack" helm chart to deploy operator and cluster components. Initial settings were: 2 select nodes, 2 storage nodes, 2 insert nodes.
2. After that i mentioned that vmcluster status is expanding. I tried to search some logs with no luck - there were some informational messages like "checking pods, reconcilling and so on". I used kubectl edit to scale vmselect to 1 to check what will happen. And found the logs about "status field couldn't be updated". Status of the cluster goes to "failed". But operator successfully removed one vmselect, so the state was really good - everything worked as expected from grafana and kubectl view.
3. I used kubectl edit to scale vmselect to 3 nodes then and found the same error about "can't update status field". Cluster status was "expanding". But operator successfully created 2 more vmselect nodes (pods are running).

I just checked another 3 clusters with vm-k8s-stack and found the same thing there - all clusters are expanding. One cluster is 1.23, second and the third are 1.24 and the last one is 1.26.

**Comment by vozerov** [comment]
Hmm... Just tried to create a new simple vmcluster on the same cluster:

```sh
apiVersion: operator.victoriametrics.com/v1beta1
kind: VMCluster
metadata:
  name: example-vmcluster
spec:
  # Add fields here
  retentionPeriod: "1"
  vmstorage:
    replicaCount: 2
  vmselect:
    replicaCount: 2
  vminsert:
    replicaCount: 2
```

And it has been created successfully:

```sh
$ kubectl -n default get vmcluster
NAME                INSERT COUNT   STORAGE COUNT   SELECT COUNT   AGE   STATUS
example-vmcluster   2              2               2              78s   operational
```

After that i modified replica count for vmselect and everything went ok:

```sh
$ kubectl -n default get vmcluster
NAME                INSERT COUNT   STORAGE COUNT   SELECT COUNT   AGE   STATUS
example-vmcluster   2              2               1              48s   operational
```

Also increased replica count to 2 back and everything was ok:

```sh
$ kubectl -n default get vmcluster
NAME                INSERT COUNT   STORAGE COUNT   SELECT COUNT   AGE    STATUS
example-vmcluster   2              2               2              119s   operational
```

I am going to simplify vmcluster resource generated by helm and try to find if it helps - will be back with some results soon.

**Comment by vozerov** [comment]
After some research i managed to find interesting thing. First of all i made a mistake when i told about "constantly expanding state". It's not constant - it changes all the time from "expanding" to "operational" and back to "expanding".

I've created the following vmcluster:

```sh
apiVersion: operator.victoriametrics.com/v1beta1
kind: VMCluster
metadata:
  annotations:
  finalizers:
  - apps.victoriametrics.com/finalizer
  labels:
    app.kubernetes.io/instance: vm
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: vm-test-cluster
    app.kubernetes.io/version: v1.91.2
    helm.sh/chart: victoria-metrics-k8s-stack-0.16.3
  name: vm-test-cluster
  namespace: default
spec:
  imagePullSecrets: []
  replicationFactor: 1
  retentionPeriod: "14"
  vminsert:
    extraArgs: {}
    image:
      tag: v1.91.2-cluster
    nodeSelector:
      usecase: workload
    replicaCount: 1
    resources: {}
  vmselect:
    cacheMountPath: /select-cache
    extraArgs:
      vmalert.proxyURL: http://vmalert-vm-victoria-metrics-k8s-stack.p2p-platform.svc:8080
    image:
      tag: v1.91.2-cluster
    nodeSelector:
      usecase: workload
    replicaCount: 1
    resources: {}
  vmstorage:
    image:
      tag: v1.91.2-cluster
    nodeSelector:
      usecase: workload
    replicaCount: 1
    resources: {}
    storage:
      volumeClaimTemplate:
        metadata:
          name: vmstorage-vm-data
        spec:
          accessModes:
            - ReadWriteOnce
          resources:
            requests:
              storage: 5G
          storageClassName: do-block-storage
    storageDataPath: /vm-data
```

And after i applied it i got the following:

```sh
 $ kubectl -n default get vmcluster
NAME              INSERT COUNT   STORAGE COUNT   SELECT COUNT   AGE   STATUS
vm-test-cluster   1              1               1              91s   operational
 $ kubectl -n default get vmcluster
NAME              INSERT COUNT   STORAGE COUNT   SELECT COUNT   AGE   STATUS
vm-test-cluster   1              1               1              91s   operational
 $ kubectl -n default get vmcluster
NAME              INSERT COUNT   STORAGE COUNT   SELECT COUNT   AGE   STATUS
vm-test-cluster   1              1               1              92s   operational
 $ kubectl -n default get vmcluster
NAME              INSERT COUNT   STORAGE COUNT   SELECT COUNT   AGE   STATUS
vm-test-cluster   1              1               1              93s   operational
 $ kubectl -n default get vmcluster
NAME              INSERT COUNT   STORAGE COUNT   SELECT COUNT   AGE    STATUS
vm-test-cluster   1              1               1              102s   expanding
 $ kubectl -n default get vmcluster
NAME              INSERT COUNT   STORAGE COUNT   SELECT COUNT   AGE    STATUS
vm-test-cluster   1              1               1              104s   expanding
 $ kubectl -n default get vmcluster
NAME              INSERT COUNT   STORAGE COUNT   SELECT COUNT   AGE    STATUS
vm-test-cluster   1              1               1              105s   expanding
 $ kubectl -n default get vmcluster
NAME              INSERT COUNT   STORAGE COUNT   SELECT COUNT   AGE    STATUS
vm-test-cluster   1              1               1              107s   operational
 $ kubectl -n default get vmcluster
NAME              INSERT COUNT   STORAGE COUNT   SELECT COUNT   AGE    STATUS
vm-test-cluster   1              1               1              110s   operational
```

So as you can see status swaps between "operational" and "expanding". And the reason for this is:

```sh
  imagePullSecrets: []
  vminsert:
    extraArgs: {}
```

So if i've removed imagePullSecrets and extraArgs from vmcluster. So now it looks like:

```sh
apiVersion: operator.victoriametrics.com/v1beta1
kind: VMCluster
metadata:
  annotations:
  finalizers:
  - apps.victoriametrics.com/finalizer
  labels:
    app.kubernetes.io/instance: vm
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: vm-test-cluster
    app.kubernetes.io/version: v1.91.2
    helm.sh/chart: victoria-metrics-k8s-stack-0.16.3
  name: vm-test-cluster
  namespace: default
spec:
  replicationFactor: 1
  retentionPeriod: "14"
  vminsert:
    image:
      tag: v1.91.2-cluster
    nodeSelector:
      usecase: workload
    replicaCount: 1
    resources: {}
  vmselect:
    cacheMountPath: /select-cache
    extraArgs:
      vmalert.proxyURL: http://vmalert-vm-victoria-metrics-k8s-stack.p2p-platform.svc:8080
    image:
      tag: v1.91.2-cluster
    nodeSelector:
      usecase: workload
    replicaCount: 1
    resources: {}
    storage:
      volumeClaimTemplate:
        metadata: {}
        spec:
          resources:
            requests:
              storage: 1Gi
        status: {}
  vmstorage:
    image:
      tag: v1.91.2-cluster
    nodeSelector:
      usecase: workload
    replicaCount: 1
    resources: {}
    storage:
      volumeClaimTemplate:
        metadata: {}
        spec:
          resources:
            requests:
              storage: 5Gi
        status: {}
    storageDataPath: /vm-data
```

Diff is:

```sh
$ diff vmcluster.yaml vmcluster-orig.yaml
15a16
>   imagePullSecrets: []
18a20
>     extraArgs: {}
```

And it starts working as expected - no swaps, constant "operational" state. And no more resourceRevision changes. Also, I've modified existing vmcluster resources on all clusters and it helped - no more swaps.

Still don't know the root reason for this, but definitely know that it relates to null imagePullSecrets and extraArgs.

**Comment by Haleygo** [comment]
@vozerov Thank you so much for such detail report, I think I know what's going wrong, will make a fix very soon.
In the meantime, you can avoid this bug by not define null imagePullSecrets&extraArgs.

**Comment by f41gh7** [comment]
Fixed at [v0.35.0](https://github.com/VictoriaMetrics/operator/releases/tag/v0.35.0) release

**Comment by liuxuzxx** [comment]
operator: 0.55
helm: victoria-metrics-k8s-stack-0.42.0
vmagent,vminsert,vmselect,vmstorage ...etc version: v1.115.0-cluster


```bash
{"level":"error","ts":"2025-04-09T06:21:46Z","logger":"manager","msg":"Reconciler error","controller":"vmcluster","controllerGroup":"[operator.victoriametrics.com](https://operator.victoriametrics.com/)","controllerKind":"VMCluster","VMCluster":{"name":"vmks-victoria-metrics-k8s-stack","namespace":"monitoring"},"namespace":"monitoring","name":"vmks-victoria-metrics-k8s-stack","reconcileID":"f7c72326-1022-43d4-802d-caf2aa03a3ad","error":"failed to update object status: cannot update resource status with patch: [VMCluster.operator.victoriametrics.com](https://vmcluster.operator.victoriametrics.com/) \"vmks-victoria-metrics-k8s-stack\" is invalid: status.updateFailCount: Required value"}
```

I upgraded from helm version 0.36.0 to 0.42.0, and then this error occurred. My solution now is:

```bash
#1. uninstall vmks
linux > helm uninstall vmks -n monitoring

#2. delete all resources crds and pvc
kubectl delete crd vmalertmanagerconfigs.operator.victoriametrics.com
kubectl delete crd vmalertmanagers.operator.victoriametrics.com
kubectl delete crd vmalerts.operator.victoriametrics.com
kubectl delete crd vmauths.operator.victoriametrics.com
kubectl delete crd vmclusters.operator.victoriametrics.com
kubectl delete crd vmnodescrapes.operator.victoriametrics.com
kubectl delete crd vmpodscrapes.operator.victoriametrics.com
kubectl delete crd vmprobes.operator.victoriametrics.com
kubectl delete crd vmrules.operator.victoriametrics.com
kubectl delete crd vmscrapeconfigs.operator.victoriametrics.com
kubectl delete crd vmservicescrapes.operator.victoriametrics.com
kubectl delete crd vmsingles.operator.victoriametrics.com
kubectl delete crd vmstaticscrapes.operator.victoriametrics.com
kubectl delete crd vmusers.operator.victoriametrics.com


....etcd pvc

#3. reinstall vmks
helm install vmks ./ -f ./re-values.yaml -n monitoring

#4. then is ok
linux > kubectl get vmcluster -n monitoring     
NAME                              INSERT COUNT   STORAGE COUNT   SELECT COUNT   AGE    STATUS
vmks-victoria-metrics-k8s-stack   2              2               2              6m2s   operational

```

**Comment by f41gh7** [comment]
@liuxuzxx could you please try to update operator CRDs? For k8s-stack helm based installations it's manual action. 

```
But release with CRD update can only be patched manually with kubectl. Since helm does not perform a CRD update, we recommend that you always perform this when updating the helm-charts version:

# 1. check the changes in CRD
$ helm show crds vm/victoria-metrics-k8s-stack --version [YOUR_CHART_VERSION] | kubectl diff -f -

# 2. apply the changes (update CRD)
$ helm show crds vm/victoria-metrics-k8s-stack --version [YOUR_CHART_VERSION] | kubectl apply -f - --server-side
```

https://github.com/VictoriaMetrics/helm-charts/tree/master/charts/victoria-metrics-k8s-stack#upgrade-guide

**Comment by liuxuzxx** [comment]
The result is:
```bash
linux > kubectl get vmcluster -n monitoring
NAME                              INSERT COUNT   STORAGE COUNT   SELECT COUNT   AGE   STATUS
vmks-victoria-metrics-k8s-stack   2              2               2              17m 
```

**Comment by liuxuzxx** [comment]
> [@liuxuzxx](https://github.com/liuxuzxx) could you please try to update operator CRDs? For k8s-stack helm based installations it's manual action.
> 
> ```
> But release with CRD update can only be patched manually with kubectl. Since helm does not perform a CRD update, we recommend that you always perform this when updating the helm-charts version:
> 
> # 1. check the changes in CRD
> $ helm show crds vm/victoria-metrics-k8s-stack --version [YOUR_CHART_VERSION] | kubectl diff -f -
> 
> # 2. apply the changes (update CRD)
> $ helm show crds vm/victoria-metrics-k8s-stack --version [YOUR_CHART_VERSION] | kubectl apply -f - --server-side
> ```
> 
> https://github.com/VictoriaMetrics/helm-charts/tree/master/charts/victoria-metrics-k8s-stack#upgrade-guide

Thanks,I didn't update it this way.

I just executed the following command:
```bash
linux > /media/liuxu/data/component/victoria-metrics/victoria-metrics-k8s-stack-0.42.0   feature/support-new-version-lx  helm install vmks ./ -f ./re-values.yaml -n monitoring 
```

and my re-values.yaml is

```yaml
global:
  cluster:
    dnsDomain: idc.local.
victoria-metrics-operator:
  image:
    registry: xwharbor.wxchina.com/cpaas
  crds:
    cleanup:
      image:
        repository: xwharbor.wxchina.com/cpaas/bitnami/kubectl
  operator:
    disable_prometheus_converter: true
defaultDashboards:
  enabled: false
  grafanaOperator:
    enabled: false
defaultRules:
  create: false

#vmsingle模式直接enabled:false，部署成vmcluster模式
vmsingle:
  enabled: false
vmcluster:
  enabled: true
  spec:
    vmstorage:
      image:
        repository: xwharbor.wxchina.com/cpaas/victoriametrics/vmstorage
      storage:
        volumeClaimTemplate:
          spec:
            storageClassName: nfs-client
            resources:
              requests:
                storage: 100Gi
      resources:
        requests:
          cpu: 500m
          memory: 1024Mi
        limits:
          cpu: 3000m
          memory: 4096Mi
    vmselect:
      image:
        repository: xwharbor.wxchina.com/cpaas/victoriametrics/vmselect
      storage:
        volumeClaimTemplate:
          spec:
            storageClassName: nfs-client
            resources:
              requests:
                storage: 10Gi
    vminsert:
      image:
        repository: xwharbor.wxchina.com/cpaas/victoriametrics/vminsert

vmalert:
  spec:
    image:
      repository: xwharbor.wxchina.com/cpaas/victoriametrics/vmalert
    configReloaderImageTag: xwharbor.wxchina.com/cpaas/jimmidyson/configmap-reload:v0.3.0
vmagent:
  enabled: true
  spec:
    replicaCount: 3
    image:
      repository: xwharbor.wxchina.com/cpaas/victoriametrics/vmagent
    resources:
      requests:
        cpu: 100m
        memory: 200Mi
      limits:
        cpu: 2000m
        memory: 2048Mi
      

alertmanager:
  spec:
    image:
      repository: xwharbor.wxchina.com/cpaas/prom/alertmanager
    configReloaderImageTag: xwharbor.wxchina.com/cpaas/jimmidyson/configmap-reload:v0.3.0

grafana:
  enabled: false
prometheus-node-exporter:
  enabled: false
kube-state-metrics:
  enabled: false
kubelet:
  enabled: false
kubeApiServer:
  enabled: false
kubeControllerManager:
  enabled: false
coreDns:
  enabled: false
kubeEtcd:
  enabled: false
kubeScheduler:
  enabled: false
```

**Comment by liuxuzxx** [comment]
![Image](https://github.com/user-attachments/assets/78674599-e7c3-445f-b933-514f8e80cc70)

I have another question. I saw the indicator vm_app_start_timestamp. This indicator seems to be the time when the vm application is started, but I saw that there are breakpoints sometimes. I checked the pod later and did not find any pod restart. What is the reason for this?

At that time, I was doing a stress test of the vm cluster mode. The collection speed was about 45w-50w/s. I don't know if this will affect this.
